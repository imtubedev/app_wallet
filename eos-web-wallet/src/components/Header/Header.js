import React from "react";
import { Link } from "react-router-dom";
import css from "./styles.module.scss";
import cx from "classnames";
import withProfile from "../../containers/profile/index";
import withUserId from "../../containers/current-user/index";

const renderHeaderUser = (
  userId,
  imageUrl,
  displayName,
  location,
  onLogout
) => (
  <div className={css.user_info}>
    <div className="has-text-right is-hidden-mobile">
      <div className={css.user_meta}>
        <h4 className={cx("title is-4 is-spaced", css.title)}>
          Hi, {displayName || userId}
        </h4>
        <p className={cx("subtitle is-6", css.subtitle)}>
          <Link to="/profile">{location || "Customize your profile"}</Link> |{" "}
          <span> </span>
          <span onClick={onLogout} className="icon-logout" />
        </p>
      </div>
    </div>

    <Link className={css.settings} to="/profile">
      <span className="icon-settings" />
    </Link>

    <figure className={cx("image", css.profile_thumbnail)}>
      <img
        src={imageUrl === "" ? "/images/user.png" : imageUrl}
        alt={displayName || userId}
      />
    </figure>
  </div>
);

const Header = ({
  children,
  isAuthenticated,
  onClick,
  onLogout,
  userId,
  userProfile: { imageUrl = "/images/user.png", displayName, location },
  ...props
}) => (
  <header className={cx("hero", css.header)}>
    <div className="level is-mobile">
      <div className="level-left">
        <div className="control level-item is-hidden-tablet">
          <button
            className={cx("button is-primary u-mr1", css.menu_toggle)}
            onClick={onClick}
          >
            <span className="icon-menu" />
          </button>
        </div>

        <div className="level-item">
          <Link to="/">
            <img
              alt=""
              className={cx("image", css.logo)}
              src="/images/logo.svg"
            />
          </Link>
        </div>
      </div>

      {isAuthenticated
        ? renderHeaderUser(userId, imageUrl, displayName, location, onLogout)
        : null}
    </div>
  </header>
);

export default withUserId(withProfile(Header));
