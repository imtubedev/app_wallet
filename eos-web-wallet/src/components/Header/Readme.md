A Header example:

```js
const { Provider } = require( "react-redux");
const { Router } = require("react-router-dom");
const { createBrowserHistory } = require("history");
const store = {
  getState: () => ({
      user: {
        isAuthenticated: true,
        email: "here@there.com"
      },
      profile: {
        profile: {
          displayName: "Here N. There",
          location: "Anytown, USA"
        }
      }
    }),
  dispatch: _ => _,
  subscribe: _ => _
};

<Provider store={store}>
  <Router history={createBrowserHistory()}>
    <Header />
  </Router>
</Provider>
```
