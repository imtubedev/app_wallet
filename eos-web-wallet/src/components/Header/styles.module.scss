@import "../../styles/custom_variables.scss";

.header {
	background-color: $grey-dark;
	position: relative;
	padding: 1em;
	height: 90px;
	@include mobile {
		padding: 0;
		height: 50px;
	}
	.menu_toggle {
		width: 50px;
		height: 50px;
		border-radius: 0;
		span {
			font-size: 1.5rem;
		}
	}
	.logo {
		height: 55px;
		@include mobile {
			height: 30px;
		}
	}
	.user_info {
		position: absolute;
		top: 1rem;
		right: 1rem;
		width: 80%;
		@include mobile {
			right: .5rem;
		}
		.user_meta {
			position: absolute;
			right: 110px;
			top: 0;
			> * {
				word-break: normal;
			}
			.title {
				margin: .25rem 0 0 0;
			}
			.subtitle, a {
				color: $secondary;
				margin-bottom: 0;
				span {
					color: $primary;
				}
			}
		}
		.settings {
			position: absolute;
			right: 75px;
			top: 55px;
			border: 5px solid $primary;
			border-radius: 15px;
			width: 30px;
			height: 30px;
			background-color: white;
			z-index: 3;
			@include mobile {
				border: 3px solid $primary;
				width: 25px;
				height: 25px;
				top: 20px;
				right: 45px;
			}
			span {
				color: $grey-dark;
				position: absolute;
				top: -1px;
				left: 3px;
				font-size: .9rem;
				@include mobile {
					font-size: .875rem;
				}
			}
		}
		.profile_thumbnail {
			position: absolute;
			right: 0;
			top: 0;
			z-index: 2;
			@include mobile {
				top: -25px;
			}
			img {
				border-radius: 50px;
				border: 8px solid $orange;
				width: 100px;
				height: 100px;
				@include mobile {
					border-radius: 35px;
					border: 5px solid $orange;
					width: 70px;
					height: 70px;
				}
			}
		}
	}
}