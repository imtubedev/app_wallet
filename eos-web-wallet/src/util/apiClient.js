// import moment from "moment";
import { merge, uniq, map, flatMap } from "lodash";
import { Buffer } from "buffer";
import eos from "eosjs";

const { ecc, Fcbuffer } = eos.modules;

// Required at front of all transactions.
// NOTE: This is a default of 64 '0's, even though it's not implemented in eosd yet
const chainId = new Buffer("00".repeat(32), "hex");

const eosEndpoint =
  process.env.NODE_ENV === "test"
    ? process.env.REACT_APP_EOSD_TESTS_URI
    : process.env.REACT_APP_EOSD_URI;

const apiEndpoint =
  process.env.NODE_ENV === "test"
    ? process.env.REACT_APP_API_TESTS_URI
    : process.env.REACT_APP_API_URI;

const authHeader = {
  Authorization: `Bearer ${process.env.REACT_APP_API_KEY}`
};

const defaultOptions = {
  method: "GET",
  headers: {
    "Content-type": "application/json"
  }
};

const defaultAPIOptions = {
  method: "GET",
  headers: {
    "Content-type": "application/json",
    ...authHeader
  }
};
// Grab the correct, sorted scope from the list of messages
// intended for a transaction
const getScope = messages =>
  uniq(
    flatMap(messages, msg => [
      msg.code,
      ...map(msg.authorization, auth => auth.account)
    ])
  ).sort();

const rejectBadResponse = (response: Response): mixed =>
  response.ok || Promise.reject(response);

const request = async (
  path: string,
  options: mixed = {},
  mergeDefaults: boolean = true
) => {
  const opts = mergeDefaults ? merge({}, defaultOptions, options) : options;
  try {
    const resp = await fetch(path, opts);
    console.log("<resp>", resp);
    await rejectBadResponse(resp);
    // only attempt to parse JSON if proper content-type
    if (resp.headers && resp.headers.get("content-type").indexOf("json") > -1)
      return resp.json();
    return resp.text();
  } catch (error) {
    console.log(error);
    if (typeof error === "string") {
      return Promise.reject(error);
    }
    if (error instanceof Error) {
      return Promise.reject(error);
    }
    const data =
      error.headers && error.headers.get("content-type").indexOf("json") > -1
        ? await error.json()
        : await error.text();
    return Promise.reject(data);
  }
};

const configureClient = (target, opts = {}) => {
  if (!opts.keyProvider)
    throw new Error(`api client connection needs a keyProvider`);

  const eosNode = eos.Testnet({
    keyProvider: opts.keyProvider,
    httpEndpoint: eosEndpoint,
    debug: process.env.NODE_ENV !== "production",
    broadcast: false
  });

  return Object.assign(target, {
    keyProvider: opts.keyProvider,
    eos: eosNode
  });
};

class APIClient {
  constructor(opts) {
    configureClient(this, opts);
    this.structs = this.eos.fc.structs;
  }

  setKeyProvider(keyProvider) {
    this.keyProvider = keyProvider;
    console.log("--> set keyProvider:", this.keyProvider);
  }

  async get(path, opts = {}) {
    const options = Object.assign({}, defaultAPIOptions, opts);
    return await request(`${apiEndpoint}${path}`, options);
  }

  async post(path, data, opts = {}) {
    const options = Object.assign({}, defaultAPIOptions, opts, {
      method: "POST",
      headers: authHeader,
      body: JSON.stringify(data)
    });
    return await request(`${apiEndpoint}${path}`, options);
  }

  async postTransaction(path, data, opts = {}) {
    let { scope, messages } = data;

    if (!messages) {
      throw new Error("APIClient.post() data missing `messages` property");
    }
    if (!Array.isArray(messages)) {
      throw new Error(
        "APIClient.post() messages should be an array of one or more messages"
      );
    }
    if (!scope) {
      // Attempt to get scope from messages
      scope = getScope(messages);
    }
    // ensure scope is unique and sorted ascending alphabetically
    scope = uniq(scope).sort();

    // build a public->private key mapping
    const keyMap = this.keyProvider.reduce((map, wif) => {
      map[ecc.privateToPublic(wif)] = wif;
      return map;
    }, {});

    try {
      // Get info on head on chain
      console.log(eosEndpoint);
      const info = await request(`${eosEndpoint}/v1/chain/get_info`);
      const head = new Date(`${info.head_block_time}Z`);
      head.setSeconds(head.getSeconds() + 60);
      const expr = head.toISOString().split(".")[0];
      // const expr = moment(new Date(`${info.head_block_time}Z`))
      //   .add(60, "seconds")
      //   .toISOString()
      //   .split(".")[0];

      // Build transaction from data
      const transaction = {
        ref_block_num: info.head_block_num & 0xffff,
        ref_block_prefix: new Buffer(info.head_block_id, "hex").readUInt32LE(8),
        expiration: expr,
        scope,
        read_scope: [],
        messages
      };
      // construct a buffer of the transaction to sign
      const trBuffer = Buffer.concat([
        chainId,
        Fcbuffer.toBuffer(this.structs.transaction, transaction)
      ]);

      const { required_keys } = await this.eos.getRequiredKeys(
        transaction,
        Object.keys(keyMap)
      );

      const signatures = [];
      (required_keys || []).forEach(pubKey => {
        if (!keyMap[pubKey]) {
          throw new Error(
            `APIClient: missing necessary private key for signing, needs private key for ${
              pubKey
            }`
          );
        }
        signatures.push(ecc.sign(trBuffer, keyMap[pubKey]));
      });

      transaction.signatures = signatures;

      return await this.post(path, transaction, opts);
    } catch (err) {
      console.error("fetch error:", err);
      return Promise.reject(err.statusText || err);
    }
  }
}

// Export singleton instance
export let apiClient = new APIClient({ keyProvider: [] });

// Export class
export default APIClient;
