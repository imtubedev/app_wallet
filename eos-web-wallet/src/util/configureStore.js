import { createStore, compose, combineReducers, applyMiddleware } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import { reducer as form } from "redux-form";
import { routerMiddleware } from "react-router-redux";
import { routerReducer } from "react-router-redux";
import app from "redux-modules/app";
import user from "redux-modules/user";
import usersList from "redux-modules/users-list";
import eosBalances from "redux-modules/eos-balances";
import notification from "redux-modules/notifications";
import transactions from "redux-modules/transactions";
import profile from "redux-modules/profile";
import { createMemoryHistory } from "history";
import middlewares from "middleware";
import sessionStorage from "redux-persist/lib/storage/session";
import { rehydrateAccounts } from "../middleware/account-persist/account-persist-actions";
import { selectWalletUserAuthenticated } from "../redux-modules/user/user-selectors";

let store;
let persistor;

const reducers = {
  app,
  form: persistReducer(
    {
      key: "forms",
      storage: sessionStorage
    },
    form
  ),
  user,
  usersList,
  eosBalances,
  notification,
  profile,
  transactions,
  routing: routerReducer
};

export const configureStore = (
  preloadState,
  history = createMemoryHistory()
) => {
  /* eslint-disable-next-line no-underscore-dangle */
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(...middlewares, routerMiddleware(history))
  );

  store = createStore(combineReducers(reducers), preloadState, enhancers);
  store.reducers = reducers;
  persistor = persistStore(store, undefined, () => {
    if (selectWalletUserAuthenticated(store.getState())) {
      store.dispatch(rehydrateAccounts());
    }
  });

  return { store, persistor };
};

export const addReducers = reducers => {
  let added = false;
  reducers.forEach(({ name, reducer }) => {
    if (!store.reducers[name]) {
      store.reducers = {
        ...store.reducers,
        [name]: reducer
      };
      added = true;
    }
  });

  if (added) {
    console.log(JSON.stringify(store.reducers));
    store.replaceReducer(combineReducers(store.reducers));
    persistor.persist();
  }
};

export const removeReducers = reducerNames => {
  let removed = false;
  reducerNames.forEach(reducerName => {
    if (store.reducers[reducerName]) {
      Reflect.deleteProperty(store.reducers, reducerName);
      removed = true;
    }
  });

  if (removed) {
    console.log(JSON.stringify(store.reducers));
    store.replaceReducer(combineReducers(store.reducers));
  }
};
