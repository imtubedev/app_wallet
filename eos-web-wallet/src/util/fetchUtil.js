//@flow
/* global fetch */

export const rejectBadResponse = (response: Response): mixed =>
  response.ok || Promise.reject(response);

const defaultAppOptions = {
  method: "GET",
  credentials: "include",
  headers: {
    "Content-Type": "application/json"
  }
};

export const appRequest = async (path: string, options: mixed = {}) => {
  try {
    const resp = await fetch(path, {
      ...defaultAppOptions,
      ...options
    });
    await rejectBadResponse(resp);
    return resp.json();
  } catch (error) {
    return Promise.reject(await error.json());
  }
};

const apiHost: string = process.env.REACT_APP_API_URI || "";
const apiKey: string = process.env.REACT_APP_API_KEY || "";
const defaultApiOptions = {
  method: "GET",
  headers: {
    "Content-type": "application/json",
    Authorization: `Bearer ${apiKey}`
  }
};

export const apiRequest = async (path: string, options: mixed = {}): any => {
  let resp;
  try {
    resp = await fetch(`${apiHost}${path}`, {
      ...defaultApiOptions,
      ...options
    });
    await rejectBadResponse(resp);
    // just grab as text (rather than check response headers for content type)
    let text = await resp.text();
    try {
      // attempt to parse as JSON
      return JSON.parse(text);
    } catch (error) {
      // Not json, format and return
      return {
        status: resp.status,
        text: resp.statusText,
        message: text
      };
    }
  } catch (error) {
    return Promise.reject(await error.json());
  }
};
