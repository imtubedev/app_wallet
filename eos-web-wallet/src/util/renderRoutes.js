import React from 'react';
import Switch from 'react-router/Switch';
import Route from 'react-router/Route';
import { withRouter } from 'react-router-dom';

const renderRoutes = (routes, extraProps = {}, switchProps = {}) => (
  routes ? (
    <Switch {...switchProps}>
      {routes.map((route, i) => {
        const RouteComponent = withRouter(route.component);
        return (
          <Route
            key={route.key || i}
            path={route.path}
            exact={route.exact}
            strict={route.strict}
            render={props => <RouteComponent {...props} {...extraProps} route={route} routes={route.routes} />}
          />);
      })}
    </Switch>
  ) : null
);

export default renderRoutes;
