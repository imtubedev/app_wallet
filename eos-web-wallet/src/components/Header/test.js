import * as React from "react";
import renderer from "react-test-renderer";
import { StaticRouter } from "react-router";
import Header from "./";
import { Provider } from "react-redux";
import { configureStore } from "../../util/configureStore";

describe("<Header />", () => {
  it("renders without crashing", () => {
    const staticContext = {};
    const { store } = configureStore();

    const tree = renderer
      .create(
        <StaticRouter location="/" context={staticContext}>
          <Provider store={store}>
            <Header />
          </Provider>
        </StaticRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
