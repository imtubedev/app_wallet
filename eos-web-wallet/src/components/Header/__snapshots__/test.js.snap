// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders without crashing 1`] = `
<header
  className="hero"
>
  <div
    className="level is-mobile"
  >
    <div
      className="level-left"
    >
      <div
        className="control level-item is-hidden-tablet"
      >
        <button
          className="button is-primary u-mr1"
          onClick={undefined}
        >
          <span
            className="icon-menu"
          />
        </button>
      </div>
      <div
        className="level-item"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          <img
            alt=""
            className="image"
            src="/images/logo.svg"
          />
        </a>
      </div>
    </div>
  </div>
</header>
`;
